interop_binaries/src/bin/janus_interop_collector.rs

```diff
@@ -0,0 +1,503 @@
+        AggregateShare, Collector, Vdaf,
+    },
+};
+use reqwest::{
+    header::{CONTENT_TYPE, LOCATION},
+    Url,
+};
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    net::{Ipv4Addr, SocketAddr},
+    sync::Arc,
+};
+use tokio::sync::Mutex;
+use warp::{hyper::StatusCode, reply::Response, Filter, Reply};
+
+static DAP_AUTH_TOKEN: &str = "DAP-Auth-Token";
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct AddTaskRequest {
+    task_id: String,
+    leader: String,
+    vdaf: VdafObject,
+    collector_authentication_token: String,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct AddTaskResponse {
+    status: &'static str,
+    #[serde(default)]
+    error: Option<String>,
+    collector_hpke_config: Option<String>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct CollectStartRequest {
+    task_id: String,
+    agg_param: String,
+    batch_interval_start: u64,
+    batch_interval_duration: u64,
+}
+
+#[derive(Debug, Serialize)]
+struct CollectStartResponse {
+    status: &'static str,
+    #[serde(default)]
+    error: Option<String>,
+    #[serde(default)]
+    handle: Option<String>,
+}
+
+#[derive(Debug, Deserialize)]
+struct CollectPollRequest {
+    handle: String,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(untagged)]
+enum AggregationResult {
+    Number(u64),
+    NumberArray(Vec<u64>),
+}
+
+#[derive(Debug, Serialize)]
+struct CollectPollResponse {
+    status: &'static str,
+    #[serde(default)]
+    error: Option<String>,
+    #[serde(default)]
+    result: Option<AggregationResult>,
+}
+
+struct TaskState {
+    private_key: HpkePrivateKey,
+    hpke_config: HpkeConfig,
+    leader_url: Url,
+    vdaf: VdafObject,
+    auth_token: String,
+}
+
+/// A collect job handle.
+#[derive(Clone, PartialEq, Eq, Hash)]
+struct Handle(String);
+
+impl Handle {
+    fn generate() -> Handle {
+        let randomness = rand::random::<[u8; 32]>();
+        Handle(base64::encode_config(randomness, URL_SAFE_NO_PAD))
+    }
+}
+
+struct CollectJobState {
+    task_id: TaskId,
+    url: Url,
+    batch_interval: Interval,
+    agg_param: Vec<u8>,
+}
+
+async fn handle_add_task(
+    tasks: &Mutex<HashMap<TaskId, TaskState>>,
+    request: AddTaskRequest,
+) -> anyhow::Result<HpkeConfig> {
+    let task_id_bytes = base64::decode_config(request.task_id, base64::URL_SAFE_NO_PAD)
+        .context("invalid base64url content in \"taskId\"")?;
+    let task_id = TaskId::get_decoded(&task_id_bytes).context("invalid length of TaskId")?;
+
+    let mut tasks_guard = tasks.lock().await;
+    let entry = tasks_guard.entry(task_id);
+    if let Entry::Occupied(_) = &entry {
+        return Err(anyhow::anyhow!("cannot add a task with a duplicate ID"));
+    }
+
+    let (hpke_config, private_key) = generate_hpke_config_and_private_key(
+        HpkeConfigId::from(0u8),
```

Comment:
If `handle_add_task` is called multiple times, won't the aggregator end up with multiple HPKE configs with the same ID? Is the notion that HPKE configs only need to be unique in the scope of a task?

Reply:
I'm not aware of any semantics for this ID, beyond that it gets copied into ciphertext structs, but intuitively I figure that would have to be only unique per task. Aggregators can definitely use different HPKE configs for different tasks, so it stands to reason collectors could too. I imagine a big enough aggregator or collector could run out of configuration identifier space with many tasks if they tried to not reuse any HpkeConfigId across tasks.

Reply:
Oh, right, this is only the collector. This should be OK for now, but might become a problem in the future when DAP revisits how the collector's HPKE config is distributed (baking it into the task makes key rotation hard).